src/lib.rs

```diff
@@ -1,38 +1,54 @@
-pub mod args;
-pub mod commands;
-pub mod env;
-pub mod error;
-
-use std::{ffi::OsString, path::Path};
+use std::ffi::OsString;
+use std::path::Path;
 
 use clap::Parser;
-
-use args::Args;
-use commands::{nsec3hash::Nsec3Hash, LdnsCommand};
+use commands::{key2ds::Key2ds, nsec3hash::Nsec3Hash, LdnsCommand};
+use env::Env;
 use error::Error;
 
-pub fn parse_args<I: IntoIterator<Item = OsString>, T: Fn() -> I>(
-    args_provider: T,
-) -> Result<Args, Error> {
-    try_ldns_compatibility(args_provider()).or_else(|_| {
-        Args::try_parse_from(args_provider()).map_err(|err| Error::new(err.to_string().as_str()))
-    })
-}
+pub use self::args::Args;
+
+pub mod args;
+pub mod commands;
+pub mod env;
+pub mod error;
 
-pub fn try_ldns_compatibility<I: IntoIterator<Item = OsString>>(args: I) -> Result<Args, Error> {
+pub fn try_ldns_compatibility<I: IntoIterator<Item = OsString>>(
+    args: I,
+) -> Result<Option<Args>, Error> {
     let mut args_iter = args.into_iter();
-    let binary_path = args_iter
-        .next()
-        .ok_or::<Error>("Missing binary name".into())?;
+    let binary_path = args_iter.next().ok_or("Missing binary name")?;
 
     let binary_name = Path::new(&binary_path)
         .file_name()
-        .ok_or::<Error>("Missing binary file name".into())?
+        .ok_or("Missing binary file name")?
         .to_str()
-        .ok_or("Binary file name is not valid Unicode")?;
+        .ok_or("Binary file name is not valid unicode")?;
 
-    match binary_name {
+    let res = match binary_name {
+        "ldns-key2ds" => Key2ds::parse_ldns_args(args_iter),
         "ldns-nsec3-hash" => Nsec3Hash::parse_ldns_args(args_iter),
-        _ => Err("Unrecognised LDNS command".into()),
+        _ => return Ok(None),
```

Comment:
This should presumably be `Err` rather than `Ok(None)`.

Reply:
I was using:
- `Ok(Some(args))`: correctly parsed ldns command
- `Ok(None)`: no ldns command matched
- `Err(err)`: an ldns command failed.

If we return an error, we get into trouble because we need to recover (i.e. call clap parsing). We could do `Option<Result<_, _>>` instead of `Result<Option<_>, _>` but that will just lead to `.transpose()?` somewhere else.

Reply:
See: https://github.com/NLnetLabs/dnst/pull/33